src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,369 @@
+  std::function<void()> operation_;
+};
+
+/**
+ * The client makes RPC calls through the provided underlying grpc client.
+ * If the call goes through, the user provided callback is invoked.
+ * If the call fails due to transient network error, it is added to a retry queue.
+ * The client waits for the grpc channel reconnection to resend the requests.
+ * If the total number of request bytes in the queue exceeds max_pending_requests_bytes,
+ * the thread is blocked until some requests are resent.
+ * If a call's timeout_ms reaches during retry, its callback is called with
+ * Status::TimedOut. If the whole client does not reconnect within
+ * server_unavailable_timeout_seconds, server_unavailable_timeout_callback is invoked.
+ * When all callers of the client release the shared_ptr of the client, the client
+ * destructor is called and the client is shut down.
+ */
+class RetryableGrpcClient : public std::enable_shared_from_this<RetryableGrpcClient> {
+ public:
+  static std::shared_ptr<RetryableGrpcClient> Create(
+      std::shared_ptr<grpc::Channel> channel,
+      instrumented_io_context &io_context,
+      uint64_t max_pending_requests_bytes,
+      uint64_t check_channel_status_interval_milliseconds,
+      uint64_t server_unavailable_timeout_seconds,
+      std::function<void()> server_unavailable_timeout_callback,
+      const std::string &server_name) {
+    auto retryable_grpc_client = std::shared_ptr<RetryableGrpcClient>(
+        new RetryableGrpcClient(channel,
+                                io_context,
+                                max_pending_requests_bytes,
+                                check_channel_status_interval_milliseconds,
+                                server_unavailable_timeout_seconds,
+                                server_unavailable_timeout_callback,
+                                server_name));
+    return retryable_grpc_client;
+  }
+
+  RetryableGrpcClient(const RetryableGrpcClient &) = delete;
+  RetryableGrpcClient &operator=(const RetryableGrpcClient &) = delete;
+
+  template <typename Service, typename Request, typename Reply>
+  void CallMethod(PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+                  GrpcClient<Service> &grpc_client,
+                  const std::string &call_name,
+                  const Request &request,
+                  const ClientCallback<Reply> &callback,
+                  const int64_t timeout_ms) {
+    auto executor = new Executor(
+        [callback](const ray::Status &status) { callback(status, Reply()); });
+    std::weak_ptr<RetryableGrpcClient> weak_self = shared_from_this();
+    auto operation_callback = [weak_self, request, callback, executor, timeout_ms](
+                                  const ray::Status &status, Reply &&reply) {
+      auto self = weak_self.lock();
+      if (status.ok() || !IsGrpcRetryableStatus(status) || !self) {
+        callback(status, std::move(reply));
+        delete executor;
```

Comment:
Can we return early?
Ref: https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

- clang-tidy issues warnings on this
- It's hard to read the code segment with such long indentation